libretroshare/src/services/p3gxsforums.cc

```diff
@@ -192,221 +192,202 @@ RsSerialiser* p3GxsForums::setupSerialiser()
 
 void p3GxsForums::notifyChanges(std::vector<RsGxsNotify *> &changes)
 {
-#ifdef GXSFORUMS_DEBUG
-	std::cerr << "p3GxsForums::notifyChanges() : " << changes.size() << "changes to notify" << std::endl;
-#endif
+	RS_DBG2(changes.size(), " changes to notify");
 
 	std::vector<RsGxsNotify *>::iterator it;
 	for(it = changes.begin(); it != changes.end(); ++it)
 	{
-		RsGxsMsgChange *msgChange = dynamic_cast<RsGxsMsgChange *>(*it);
-
-		if (msgChange)
+		rs_view_ptr<RsGxsNotify> gxsChange = *it;
+		switch(gxsChange->getType())
 		{
-			if (msgChange->getType() == RsGxsNotify::TYPE_RECEIVED_NEW || msgChange->getType() == RsGxsNotify::TYPE_PUBLISHED) /* message received */
-				if (rsEvents)
+		case RsGxsNotify::TYPE_RECEIVED_NEW: // [[fallthrough]]
+		case RsGxsNotify::TYPE_PUBLISHED:
+		{
+			rs_view_ptr<RsGxsMsgChange> msgChange =
```

Comment:
it's a bit inconsistent to dynamic_cast after entering the types of changes, given that MsgChange and GrpChange do not have the same subtypes. I understand it factors the code and maybe makes it a little smaller, but its harder to read now.

Reply:
`dynamic_cast` is very costly compared to an almost costless switch on an integer, so we should do that only in strictly necessary cases. also I think it would be much better to have more detailed RsGxsNotify enum type so each case can get only one object type, so we do just one cast per case in the eventuality it is strictly needed, so were I perceive the inconsistency is that the same `GxsNotify::TYPE_XXX` we can get different actual object types, or did I misunderstood the code?

Reply:
The cost for sure would be a problem if you're in a loop. I'm not so sure it has any impact here. But surely, the different way of treating GxsNotify objects in forums (as opposed to all other GXS services) raises questions when reading the code.

Reply:
Actually that code is inside a for loop ;-p